### ISSUE TITLE
Add a check to ensure proper usage of interfaces

### ISSUE THREAD

**Issue by hco** [phpmd/phpmd#92]
I am looking for a check which should verify methods which have a typehint on interfaces and probably abstract classes.

Whenever a method expects an object which should implement a specific interface, only calls to methods that are implemented on that interface should be allowed.

I would really love this feature for abstract classes as well, but I'm not sure if it's actually desired by anyone else there, due to magic methods.

Example:

``` php
<?php
interface BarProvider
{
    public function getBar();
}

class Foo implements BarProvider
{
    public function getBar()
    {

    }

    public function getBaz()
    {

    }
}

public function asdf(BarProvider $barProvider)
{
    $barProvider->getBar(); // Valid
    $barProvider->getFoo(); // Should raise an error
}
```

**Comment by ravage84** [comment]
Closing after 7 years of no further demand on this.
